File: js/example_hard_data_multiple.js
const {SolverAiComputeInput, CONSTRAINT, OBJECTIVE} = require('./SolverAiComputeInput');
const SolverAiClientSetup = require('./SolverAiClientSetup');
const SolverAiClientCompute = require('./SolverAiClientCompute');

const config = require('./setup.js');


const solverAiClientSetup = 
    new SolverAiClientSetup(config.datamanagerUrl, config.token);

async function main() {
    var exitCode = 1;

    const equation_ids = [];
    const code_ids = [];
    const hard_data_ids = [];
    const soft_data_ids = [];
    let problem_id = null;

    try {

        let id = await solverAiClientSetup.postHardData(
            'hard_data_basic_1',
            `${config.data_file_folder_path}/hard_data_basic_1.csv`
        )
        hard_data_ids.push(id);

        id = await solverAiClientSetup.postHardData(
            'hard_data_basic_2',
            `${config.data_file_folder_path}/hard_data_basic_2.csv`
        )
        hard_data_ids.push(id);

        id = await solverAiClientSetup.postHardData(
            'hard_data_basic_3',
            `${config.data_file_folder_path}/hard_data_basic_3.csv`
        )
        hard_data_ids.push(id);

        problem_id = await solverAiClientSetup.postProblem(
            'Test Problem',
            equation_ids,
            code_ids,
            hard_data_ids,
            soft_data_ids
        );

        const solverAiClientCompute = new SolverAiClientCompute(config.computerUrl, config.token, problem_id);

        let [inputs, outputs] = await solverAiClientCompute.getProblemSetup();

        if (JSON.stringify(inputs) !== JSON.stringify([]) || 
            JSON.stringify(outputs) !== JSON.stringify(['C1', 'T1', 'T2', 'T3']))
        {
            throw new Error('Problem Setup JSON does not match expected value.');
        }

        let input = new SolverAiComputeInput(problem_id);
        input.addObjective("T3", OBJECTIVE.MINIMIZE);

        const results = await solverAiClientCompute.runSolver(input);

        if (results.getNumberOfResults() < 1) {
            throw new Error('Results not as expected.');
        }
        
        console.log('Test was successful!!!');

        exitCode = 0;
    } catch (error) {
        console.error(`Exception: ${error.message}`);
    } finally {
        await solverAiClientSetup.deleteAll(
            equation_ids,
            code_ids,
            hard_data_ids,
            soft_data_ids,
            problem_id
        );
    }

    return exitCode;
}

(async () => {
    const exitCode = await main();
    process.exitCode = exitCode;
})();

File: python/example_hard_data_multiple.py
from os import path

from SolverAiClientSetup import SolverAiClientSetup
from SolverAiClientCompute import SolverAiClientCompute

from setup import token, datamanagerUrl, computerUrl, \
    data_file_folder_path


def main():
    solverAiClientSetup = \
        SolverAiClientSetup(datamanagerUrl, token)

    equation_ids = list()
    code_ids = list()
    hard_data_ids = list()
    soft_data_ids = list()
    problem_id = None

    try:

        id = solverAiClientSetup.postHardData(
            'hard_data_basic_1',
            path.join(data_file_folder_path, 'hard_data_basic_1.csv')
        )
        hard_data_ids.append(id)

        id = solverAiClientSetup.postHardData(
            'hard_data_basic_2',
            path.join(data_file_folder_path, 'hard_data_basic_2.csv')
        )
        hard_data_ids.append(id)

        id = solverAiClientSetup.postHardData(
            'hard_data_basic_3',
            path.join(data_file_folder_path, 'hard_data_basic_3.csv')
        )
        hard_data_ids.append(id)

        problem_id = solverAiClientSetup.postProblem(
            'Test Problem',
            equationIds=equation_ids,
            codeIds=code_ids,
            hardIds=hard_data_ids,
            softIds=soft_data_ids
        )

        solverAiClientCompute = \
            SolverAiClientCompute(computerUrl, token, problem_id)

        problem_setup_json = solverAiClientCompute.getProblemSetup()

        expected_problem_setup_json = {
            'id': problem_id,
            'inputs': [],
            'outputs': ['C1', 'T1', 'T2', 'T3']
        }

        assert problem_setup_json == expected_problem_setup_json

        input_json = {
            "id": problem_id,
            "inputs": {
            },
            "constraints": {
            },
            "objectives": {
                "T3": {
                    "Operation": 'minimize'
                    # Operation options are:
                    # - 'minimize'
                    # - 'maximize'
                },
            }
        }

        results = solverAiClientCompute.runSolver(input_json)

        if 'Number Of Results' not in results \
                or results['Number Of Results'] < 1:
            raise Exception('Results not as expected.')

        # results should have value similar to
        #     {
        #         'Number Of Results': 1,
        #         'Objective Variable Names': "['T3']",
        #         'F0': '[3.1]',
        #         'Constraint Variable Names ': "[]",
        #         'G0': '[]',
        #         'Input Variable Names': "[]",
        #         'X0': '[]',
        #         'Output Variable Names':
        #             "['C1', 'T1', 'T2', 'T3', 'var1', 'var2', 'var3']",
        #         'Y0': "[4.1, 1.2, 2.2, 3.1, 'B', 'H', 'P']"
        #     }

        print('Test was successful!!!')

    except Exception as e:
        print(f'Exception: {str(e)}')

    finally:
        solverAiClientSetup.deleteAll(
            equationIds=equation_ids,
            codeIds=code_ids,
            hardIds=hard_data_ids,
            softIds=soft_data_ids,
            problemId=problem_id
        )


if __name__ == "__main__":
    main()
    exit(0)
